mpc_shifting_knapsack/shifting_knapsack.mpc

```diff
@@ -0,0 +1,150 @@
+    '''
+    The recursive param is to choose if solving the problem using the iterative
+    version or the recursive one.
+    '''
+    N = len(weights)
+    lwC = math.ceil(math.log2(C + 1)) + 1
+    ret = Array(1, sint)
+    ret[0] = 0
+
+    dp = Matrix(2, C + 1, sint)
+    dp.assign_all(0)
+
+    @for_range(N)
+    def _(i):
+        @for_range(C + 1)
+        def _(j):
+            dp[i % 2][j] = dp[(i + 1) % 2][j]
+
+        cyclic_shift(dp[(i + 1) % 2], C + 1 - weights[i], lwC)
+
+        @for_range(C + 1)
+        def _(j):
+            opt2 = dp[(i + 1) % 2][j] + values[i]
+            should_mov = (j >= weights[i]) * (opt2 > dp[i % 2][j])
+
+            dp[i % 2][j] = should_mov.if_else(opt2, dp[i % 2][j])
+
+        increased = dp[i % 2][C] > ret[0]
+
+        ret[0] = increased.if_else(dp[i % 2][C], ret[0])
+
+    return ret[0]
+
+
+def bitwise_xor(a, b):
+    bits_a = a.bit_decompose()
+    bits_b = b.bit_decompose()
+    bits_xor = Array(len(bits_a), sint)
+
+    for i in range(len(bits_a)):
+        bits_xor[i] = bits_a[i].bit_xor(bits_b[i])
+
+    return sint.bit_compose(bits_xor)
+
+
+def bitwise_and(a, b):
+    bits_a = a.bit_decompose()
+    bits_b = b.bit_decompose()
+    bits_and = Array(len(bits_a), sint)
+
+    for i in range(len(bits_a)):
+        bits_and[i] = bits_a[i].bit_and(bits_b[i])
+
+    return sint.bit_compose(bits_and)
+
+
+def cyclic_shift(values, K, max_log_K):
+    K = sint(K) 
+
+    @for_range(max_log_K)
+    def _(i):
+        should_mov = bitwise_and(K, (cint(1 << (max_log_K - i - 1)))) != 0
+
+        maybe_shift_by_k(should_mov, values, 1 << (max_log_K - i - 1), 0,
+                         len(values))
+
+        sK = bitwise_xor(K, cint(1 << (max_log_K - i - 1)))
+
+        K.update(should_mov.if_else(sK, K))
+
+
+def is_base_case(K, start, end, n_values):
+    # Computations for the base case
+    N = end - start
+    K = K % N
+
+    return K == 0
+
+
+def maybe_shift_by_k(enabled, values, K, i, j):
+    start = Array(1, regint)
+    end = Array(1, regint) 
+
+    start[0] = regint(i)
+    end[0] = regint(j)
+
+    K_temp = K
+    K = Array(1, regint)
+    K[0] = regint(K_temp)
+
+    K2 = Array(1, regint)
+    K2[0] = regint(1)
+
+    condition = (1 - is_base_case(K[0], start[0], end[0], len(values))) * (K2[0] != 0)
+
+    @while_do(lambda: condition)
+    def _():
+        N = end[0] - start[0]
+        K[0] = K[0] % N
+        
+        @for_range(N - K[0])
+        def _(l):
+            aux = values[start[0] + ((l + K[0]) % N)]
+            values[start[0] + ((l + K[0]) % N)] = enabled.if_else(
+                values[start[0] + l],
+                values[start[0] + ((l + K[0]) % N)]
+            )
+
+            values[start[0] + l] = enabled.if_else(aux, values[start[0] + l]) 
+
+        # Condition for next iteration
+        K2[0] = (K[0] - (N % K[0])) % K[0]
+
+        # Definition of subproblem
+        start[0] = end[0] - K[0]
+        K[0] = K2[0]
+
+        condition.update(
```

Comment:
This can be subsumed as part of the `@while_do` condition by returning `start[0]`, `end[0]` and `K2[0]`

Reply:
Using the following source code and it's corresponding version using `Array`:

```python
a = MemValue(regint(1))
b = MemValue(regint(2))
c = MemValue(regint(0))

@for_range(30000000)
def _(i):
    c.iadd(i * (a + b))

print_ln("%s", c)
```

The following results show that `MemValue` is slightly slower than `Array` for 30M iterations.

![image](https://github.com/hashcloak/flashbots-privacy/assets/13358244/d802c173-a0c1-4283-bd3b-beb82fe00279)

Reply:
Using multiple return values is not supported as mentioned in https://github.com/data61/MP-SPDZ/issues/1048. However, I  wrote the while loop in such a way that the condition is updated automatically.

Reply:
I forgot about this despite asking that issue. Resolved